File: Editorial/170_Two_Sum_III_Data_structure_design/TypeScript/1_Sorted_List.ts
class TwoSum {
    private nums: number[];
    private is_sorted: boolean;

    constructor() {
        this.nums = [];
        this.is_sorted = false;
    }

    add(number: number) {
        this.nums.push(number);
        this.is_sorted = false;
    }

    find(value: number) {
        if (!this.is_sorted) {
            this.nums.sort((a, b) => a - b);
            this.is_sorted = true;
        }
        let low = 0,
            high = this.nums.length - 1;
        while (low < high) {
            let twosum = this.nums[low] + this.nums[high];
            if (twosum < value) low += 1;
            else if (twosum > value) high -= 1;
            else return true;
        }
        return false;
    }
}

File: Editorial/0101-0200/0170_Two_Sum_III_Data_structure_design/Go/1_Sorted_List.go
type TwoSum struct {
    nums      []int
    is_sorted bool
}

func Constructor() TwoSum {
    return TwoSum{nums: []int{}, is_sorted: false}
}

func (this *TwoSum) Add(number int) {
    this.nums = append(this.nums, number)
    this.is_sorted = false
}

func (this *TwoSum) Find(value int) bool {
    if !this.is_sorted {
        sort.Ints(this.nums)
        this.is_sorted = true
    }
    low, high := 0, len(this.nums)-1
    for low < high {
        var twosum int = this.nums[low] + this.nums[high]
        if twosum < value {
            low += 1
        } else if twosum > value {
            high -= 1
        } else {
            return true
        }
    }
    return false
}
